### ISSUE TITLE
Different behavior in printing of nil value *js.Object vs. regular struct

### ISSUE THREAD

**Issue by slimsag** [gopherjs/gopherjs#625]
The way that a `nil` `*js.Object` is printed is different than the way that a `nil` user-defined struct is printed. I ran into this when attempting to use `fmt.Sprintf("%+v", someJSObject)` for test comparison purposes.

For example, this program on the playground (https://gopherjs.github.io/playground/#/MP9gVpB6AP):

```Go
package main

import (
	"fmt"

	"github.com/gopherjs/gopherjs/js"
)

type Object struct{}

func main() {
	fmt.Println("Printing a nil *Object (regular struct):")
	a := struct {
		o *Object
	}{}
	fmt.Printf("a.o: %+v\n", a.o)
	fmt.Printf("a: %+v\n", a)

	fmt.Println("\nPrinting a nil *js.Object:")
	b := struct {
		o *js.Object
	}{}
	fmt.Printf("b.o: %+v\n", b.o)
	fmt.Printf("b: %+v\n", b)
}
```

Produces:

```
Printing a nil *Object (regular struct): 
a.o: <nil> 
a: {o:<nil>} 
 
Printing a nil *js.Object: 
b.o: null 
b: {o:0x1} 
```

i.e. a `nil` `*js.Object` is printed as `null` instead of `nil` on its own, and when inside a struct it is printed as `0x1` despite being `nil`.

**Comment by shurcooL** [comment]
Do you have thoughts on the expected behavior? If so, can you post that?

**Comment by slimsag** [comment]
Sorry if I didn't make that clear enough, the expected behavior is exactly the same as a normal nil structure. i.e.

Expected:
```
Printing a nil *js.Object: 
b.o: <nil>
b: {o:<nil>} 
```

Got:

```
Printing a nil *js.Object: 
b.o: null 
b: {o:0x1} 
```

**Comment by shurcooL** [comment]
So, this is a question of how the `js` package should intersect with `fmt` package. `(*js.Object)(nil)` is a special value that represents JavaScript `null`. I think this needs a decision.

At least somewhat related are #617 and #542.

**Comment by myitcv** [comment]
Yes definitely related to #617; I've tried to update the title of that to be more useful

**Comment by slimsag** [comment]
Does `struct{o *js.Object}{}` being printed as `{o:0x1} ` also need decision? To me that seems to be clearly a bug, given the struct field `o` on its own prints as `null`. It should print as `{o:null}` today at least.

**Comment by shurcooL** [comment]
The decision I want is to know whether things (mostly `(*js.Object)(nil)`, `js.Undefined`) from `js` package are supported for printing with `fmt` at all.

If it is supported, then it's a matter of deciding the details of how.

**Comment by shurcooL** [comment]
Relevant, the current behavior is documented/explained on the [JavaScript Tips and Gotchas wiki page](https://github.com/gopherjs/gopherjs/wiki/JavaScript-Tips-and-Gotchas):

> - In structs with *js.Object fields, the object exists in two parts. The inner *js.Object part stores the fields JavaScript will see. Those fields _must_ be tagged with `js:"fieldName"`. Other untagged fields will be stored in the outer object. [Try it.](http://www.gopherjs.org/playground/#/NSnWFAx_wA)
>
>	```go
>	type T struct {
>		*js.Object
>		I int `js:"i"`
>		J int
>	}
>
>	t1 := &T{Object: js.Global.Get("Object").New(), J: 20}
>	t1.I = 10
>	fmt.Printf("%#v\n", t1) // &main.T{Object:(*js.Object)(0x1), I:10, J:20}
>	println(t1) // Console shows something similar to
>	            // { I: 0, J: 20, Object: { i: 10 } }
>	```

**Comment by myitcv** [comment]
@shurcooL 

> In structs with *js.Object fields, the object exists in two parts

This is something I had intended to move onto after #617 (although we could move it to another issue to avoid hijacking this thread). Not specifically the two sides of such objects (although I can't see a use case for that) but the means by which we declare `*js.Object`-special types. 

At the moment I think it's too subtle for the author and the reader; I would prefer:

* an explicit tag on the `*js.Object` special field (the current check requires it be the first field)
* making the `js` tags optional; they become unnecessary because of the previous point if the casing on the field (or type for anonymous fields) is correct

```go
type S1 struct {
	underlying *js.Object `jsSpecial:true` // something like this

	Name string
	age  int `js:"age"`
}
```

This was also discussed in passing in #236
